wazo_calld/plugins/parking_lots/services.py

```
@@ -0,0 +1,231 @@
+    Timeout: NotRequired[int]
+    ParkingSpace: NotRequired[str]
+    TimeoutChannel: NotRequired[str]
+    AnnounceChannel: NotRequired[str]
+
+
+class ParkingService:
+    def __init__(
+        self,
+        ami: AmidClient,
+        ari: AriClient,
+        bus: BusConsumer,
+        confd: ConfdClient,
+        notifier: ParkingNotifier,
+    ):
+        self._amid = ami
+        self._ari = ari
+        self._confd = confd
+        self._notifier = notifier
+        self._parkings = ParkingLotCache(bus, confd)
+
+    def _get_connected_channel(self, channel: Channel) -> Channel:
+        try:
+            return channel.only_connected_channel()
+        except NotEnoughChannels:
+            raise InvalidCall(channel.id, 'no parkable peer call found')
+        except TooManyChannels:
+            raise InvalidCall(channel.id, 'cannot park a conference call')
+
+    def _park_channel(
+        self,
+        parkee: Channel,
+        parking: ConfdParkingLot,
+        tenant_uuid: str,
+        preferred_slot: str | None = None,
+        timeout: int | None = None,
+    ) -> AsteriskParkedCall:
+        park_payload: ParkActionDict = {
+            'Channel': parkee.asterisk_name(),
+            'Parkinglot': f'parkinglot-{parking.id}',
+        }
+
+        if timeout:
+            # Convert to milliseconds for AMI action
+            park_payload['Timeout'] = timeout * 1000
+
+        if preferred_slot:
+            # Attempt to use slot if it can, else it wil be auto-attributed
+            park_payload['ParkingSpace'] = preferred_slot
+
+        parked_call = self.find_parked_call(tenant_uuid, parking.id, parkee.id)
+        if not parked_call:
+            callback_channel = self._get_connected_channel(parkee)
+            park_payload['TimeoutChannel'] = callback_channel.asterisk_name()
+        else:
+            # If call is already parked, preserve callback
+            park_payload['TimeoutChannel'] = parked_call.parker_dial_string
```

If call is already parked, should we just thrown an exception?
(I didn't test) Swapping feature should be in a dedicated endpoint

---

This is more related to reparking a call since it wont leave the parking.  We could throw an exception to prevent a call from being reparked, but reparking allows to update timeout for instance.  A call can also be reparked in the same space